content/blog/fp-js-higher-order-functions.md

```diff
@@ -0,0 +1,520 @@
+initial value. So, the first pass, our accumulator `productsObject` will be an empty object.
+We then return a new object, copying all properties from the accumulator over via the objet spread operator (`...`).
+The first time this runs, it will merge together two empty objects.
+Second, we create a new property on our return object, with the key set to the product id, using javascript dynamic object properties (`{[key]: value}`).
+Lastly, we set this property to an object following our backend developers structure. Everyone is happy!
+
+<!-- Implement a map with reduce to show the versatilityt -->
+
+### Reduce is versatile
+
+`reduce` is probably the most versatile of all higher order functions built in to the `Array` object. As you saw, it can be used to create numbers or objects. We can also use it to re-implement both the `map` and the `filter` function
+
+```js
+function map(arr, mapFn) {
+  return arr.reduce((accumulator, current) => {
+    return [...accumulator, mapFn(current)];
+  }, []);
+}
+
+function filter(arr, filterFn) {
+  return arr.reduce((accumulator, current) => {
+    const temp = [...accumulator];
+    if (filterFn(current)) {
+      temp.push(current);
+    }
+    return temp;
+  }, []);
+}
+```
+
+### Beware of the index!
+
+When using these build in functions, there are some caveats, especially when using map.
+As you've seen, callback functions sent to the `map` and `filter` takes the current element it's operating on as the first argument, and `reduce` takes
+an accumulator and the current element. However, there are two more arguments that are passed in, the current index and the array we are operating on (`currentIndex` and `array`)
+
+Here is the declaration for the `map` and `filter` functions.
+
+```js
+function callback(element, index, array)
+```
+
+And for the `reduce`
+
+```js
+function callback(accumulator, element, index, array)
+```
+
+Now, take for example this array
+
+```js
+const numbers = ['1', '2', '3', '4'];
+```
+
+Say you want to parse those strings to numbers. You could use `map` in combination with `parseInt` for this.
+
+```js
+const parsed = numbers.map(n => parseInt(n)); // [1, 2, 3, 4]
+```
+
+But you also want your code to be all tidy and nice, so you decide to just pass the `parseInt` directly to `map`, since the `map` callback and the
+`parseInt` functions both take a number as their first argument.
+
+```js
+const parsed = numbers.map(parseInt);
+```
+
+Now, if you run this code, you will get this not so nice looking result
+
+```js
+[1, NaN, NaN, NaN];
+```
+
+What is going on here? Well, remember the second argument to the function passed in to `map`? Yeah, a number representing the current index of the element in our array. And what is the second argument to `parseInt`? Also a number! However, this argument is the `radix`, or `base` in which to parse the number. This is where we run into trouble! Here is what's going on:
```

Comment:
**Suggestion to be extra clear:**
Well, remember the second argument to the callback function passed in to `map`

Reply:
Fixed.